### PR TITLE
[SPARK-24266][K8S][3.0] Restart the watcher when we receive a version changed from k8s

### PR DESCRIPTION
**Pull request by jkleckner**
### What changes were proposed in this pull request?

This is a straight application of #28423 onto branch-3.0

Restart the watcher when it failed with a HTTP_GONE code from the kubernetes api. Which means a resource version has changed.

For more relevant information see here: https://github.com/fabric8io/kubernetes-client/issues/1075

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

This was tested in #28423 by running spark-submit to a k8s cluster.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
cc @holdenk

**Comment by SQUIDwarrior** [comment]
We also see this problem with Spark 2.4.X so I would very much like to see this landed for both versions.

**Comment by holdenk** [comment]
Ok sounds like something to keep in mind.
Jenkins retest this please.

**Comment by jkleckner** [comment]
This comment exists merely to link this PR with PR #29496 intended for Spark 2.4 branch

**Comment by dongjoon-hyun** [comment]
Any review update, @holdenk ?

**Comment by holdenk** [comment]
Jenkins retest this please.

**Comment by holdenk** [comment]
I haven’t had a chance to look, but given it’s a backport if you have don’t feel like you need to wait for me.

**Comment by jkleckner** [comment]
Is there some way to run that expanded test information for this single failed R integration test as described in the message?

```
- Run SparkR on simple dataframe.R example *** FAILED ***
  The code passed to eventually never returned normally. Attempted 70 times over 2.0003406448333334 minutes. Last failure message: false was not true. (KubernetesSuite.scala:315)
Run completed in 12 minutes, 36 seconds.
Total number of tests run: 19
Suites: completed 2, aborted 0
Tests: succeeded 18, failed 1, canceled 0, ignored 0, pending 0
*** 1 TEST FAILED ***
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 3.0.2-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  3.625 s]
[INFO] Spark Project Tags ................................. SUCCESS [  8.191 s]
[INFO] Spark Project Local DB ............................. SUCCESS [  3.628 s]
[INFO] Spark Project Networking ........................... SUCCESS [  5.584 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [  2.920 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 10.042 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  3.774 s]
[INFO] Spark Project Core ................................. SUCCESS [02:20 min]
[INFO] Spark Project Kubernetes Integration Tests ......... FAILURE [15:49 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  18:48 min
[INFO] Finished at: 2020-08-29T17:24:33-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.scalatest:scalatest-maven-plugin:2.0.0:test (integration-test) on project spark-kubernetes-integration-tests_2.12: There are test failures -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :spark-kubernetes-integration-tests_2.12

```

**Comment by jkleckner** [comment]
Does the integration test flakiness described by @holdenk in SPARK-32354 [1] apply to this build?

[1] https://issues.apache.org/jira/browse/SPARK-32354

**Comment by holdenk** [comment]
Huh did someone re-enable the R tests?
Anyways Jenkins retest this please.

**Comment by dongjoon-hyun** [comment]
I didn't see any change for reenabling. I guess it's disabled by Jenkins environment issue and is recovered due to some reason.

**Comment by dongjoon-hyun** [comment]
BTW, cc @zhengruifeng for 3.0.1 RC3 since the K8s IT might be broken on RC3.

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by zhengruifeng** [comment]
@dongjoon-hyun  seems like it's not a blocker in the jira.

**Comment by dongjoon-hyun** [comment]
Yes. The issue itself is not a blocker. I pinged you because of the K8s integration fails consistently on SparkR test.

**Comment by jkleckner** [comment]
So is this is ready to merge to 3.0?  I guess we just missed 3.0.1.

**Comment by holdenk** [comment]
Jenkins retest this please

**Comment by dongjoon-hyun** [comment]
It seems that `SparkR` test fail.
```
KubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- All pods have the same service account by default
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Run SparkPi with env and mount secrets.
- Run PySpark on simple pi.py example
- Run PySpark with Python2 to test a pyfiles example
- Run PySpark with Python3 to test a pyfiles example
- Run PySpark with memory customization
- Run in client mode.
- Start pod creation from template
- PVs with local storage
- Launcher client dependencies
- Run SparkR on simple dataframe.R example *** FAILED ***
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala:28]: @jkleckner Apache Spark community always use absolute path. Please fix this.
> Always import packages using absolute paths (e.g. scala.util.Random) instead of relative ones (e.g. util.Random).

**Comment by holdenk** [comment]
Jenkins retest this please.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala]: Done.
Note that the original merged patch didn't do this but was fixed later on master in  527cd3fc3aac40f84ba8eee291e1a955e03f7665

**Comment by jkleckner** [comment]
Regarding *did someone re-enable the R tests?* @holdenk should they be disabled on the `branch-3.0` branch?

**Comment by holdenk** [comment]
I think they should be disabled, the Jenkins install isn't configured for testing the R version in 3.0. Go ahead and comment it out.

**Comment by jkleckner** [comment]
Looking more carefully at history and #22145 it appears that simply rebasing should fix this.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala
Comment (jkleckner) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/KubernetesClientApplication.scala]: Ok, rebase fixed integration test.  Ok to resolve this thread?

**Comment by jkleckner** [comment]
Well, rebase didn't do it.
@shaneknapp Could this problem be similar to what you saw in #22145  ?

**Comment by jkleckner** [comment]
@holdenk On reflection, I don't think this patch set should both turn off SparkR k8s testing and implement this fix.  They should probably be separate PRs if it is desirable to turn it off so that it is easily reverted.  Is it possible to merge this even with the presence of the integration test failure?

**Comment by holdenk** [comment]
I would prefer to instead disable the R tests first. You can of course submit that in another PR and ping me to review that one.

**Comment by jkleckner** [comment]
@holdenk  Ok, integration tests pass.

It was just a two word deletion...

**Comment by jkleckner** [comment]
@dongjoon-hyun Is this ok to merge?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:45]: Please revert this. `branch-3.0` doesn't have R test issue.

```
KubernetesSuite:
- Run SparkPi with no resources
- Run SparkPi with a very long application name.
- Use SparkLauncher.NO_RESOURCE
- Run SparkPi with a master URL without a scheme.
- Run SparkPi with an argument.
- Run SparkPi with custom labels, annotations, and environment variables.
- All pods have the same service account by default
- Run extraJVMOptions check on driver
- Run SparkRemoteFileTest using a remote data file
- Run SparkPi with env and mount secrets.
- Run PySpark on simple pi.py example
- Run PySpark with Python2 to test a pyfiles example
- Run PySpark with Python3 to test a pyfiles example
- Run PySpark with memory customization
- Run in client mode.
- Start pod creation from template
- PVs with local storage
- Launcher client dependencies
- Run SparkR on simple dataframe.R example
Run completed in 8 minutes, 12 seconds.
Total number of tests run: 19
Suites: completed 2, aborted 0
Tests: succeeded 19, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Hi, @@jkleckner . 

I briefly take a look. The change on `KubernetesSuite.scala` should be reverted from this PR.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala
Comment (jkleckner) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/KubernetesSuite.scala:45]: @dongjoon-hyun  We have a review loop.
@holdenk asked me to turn off the SparkR tests - see above.
And @holdenk created a story for master to make the integration tests work properly for SparkR here [1].

[1] https://issues.apache.org/jira/projects/SPARK/issues/SPARK-32354

**Comment by redsk** [comment]
@jkleckner I tried this patch in production but it does not seem to work. 

Distribution:
```
git clone https://github.com/apache/spark.git
git checkout branch-3.0
git fetch origin pull/29533/head:backport-SPARK-24266-to-branch-3.0
git checkout backport-SPARK-24266-to-branch-3.0
git checkout -b rebased-backport-SPARK-24266-to-branch-3.0
git rebase branch-3.0
export MAVEN_OPTS="-Xmx2g -XX:ReservedCodeCacheSize=1g"
./dev/make-distribution.sh --name spark-3.0-24266 --tgz -Phive -Phive-thriftserver -Pmesos -Pyarn -Pkubernetes
```

Image creation:
```
./bin/docker-image-tool.sh -r my-registry -t my-tag -n -u 0 -b java_image_tag=11-jre-slim build
./bin/docker-image-tool.sh -r my-registry -t my-tag push
```

When I execute my long-running spark application I get
```
...
20/10/22 17:30:20 WARN ExecutorPodsWatchSnapshotSource: Kubernetes client has been closed (this is expected if the application is shutting down.)
io.fabric8.kubernetes.client.KubernetesClientException: too old resource version: 1543007015 (1543067888)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onMessage(WatchConnectionManager.java:259)
	at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:323)
	at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:274)
	at okhttp3.internal.ws.RealWebSocket$2.onResponse(RealWebSocket.java:214)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:203)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
```

I'm not sure if I made a mistake or there's a problem in the patch. Thanks

**Comment by jkleckner** [comment]
> I tried this patch in production but it does not seem to work.
> 
> ...
> I'm not sure if I made a mistake or there's a problem in the patch. Thanks

Your rebase of the patch looks correct when I tried it out.

Unfortunately, there must be something in the patches to k8s code after 2.4 and before the fix #28423 was merged.

My interest in landing this patch has been to unblock #29496 due to the backporting policies for Spark 2.4 which is what we use and I don't have a setup to test this for the 3.0 branch.

@redsk if you would like to look into that, it would be helpful.

The candidate patch sets that are in the k8s can be viewed with something like this:
`git log --oneline  branch-3.0..0432379f9923768a767566e9ac5a4021cfe8d052 --  resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s`
```
0432379f99 [SPARK-24266][K8S] Restart the watcher when we receive a version changed from k8s
c8f3bd861d [SPARK-31696][K8S] Support driver service annotation in K8S
85dad37f69 [SPARK-31601][K8S] Fix spark.kubernetes.executor.podNamePrefix to work
b8ccd75524 [SPARK-29905][K8S] Improve pod lifecycle manager behavior with dynamic allocation
7699f765f5 [SPARK-31394][K8S] Adds support for Kubernetes NFS volume mounts
ed06d98044 [SPARK-25355][K8S] Add proxy user to driver if present on spark-submit
1254c88034 [SPARK-31118][K8S][DOC] Add version information to the configuration of K8S
d273a2bb0f [SPARK-20628][CORE][K8S] Start to improve Spark decommissioning & preemption support
f9f06eee98 [SPARK-30122][K8S] Support spark.kubernetes.authenticate.executor.serviceAccountName
86fdb818bf [SPARK-30715][K8S] Bump fabric8 to 4.7.1
```

@stijndehaes can you guide as to which of these might affect this backport?

**Comment by stijndehaes** [comment]
@redsk @jkleckner The error line you are seeing comes from the class `ExecutorPodsWatchSnapshotSource` this is somewhere else. I thought there was another mechanism for the executors to work with connection problems.

This also looks like driver logs? The fix here is in the spark-submit application, you should watch the logs of that i.s.o. the driver. Can you tell me if these were driver logs or spark-submit logs?

**Comment by dongjoon-hyun** [comment]
Hi, @jkleckner and all.
Is there any updates on this PR?

**Comment by jkleckner** [comment]
@dongjoon-hyun  Since I don't have a test environment to try out or observe the problem mentioned by @redsk it will have to be taken up by someone like @redsk.

I am fully on a critical path for the next many weeks and the back port to 2.4 is working fine for us so I can't spend any more time on this now.

And as mentioned by @stijndehaes , this could very well be a separate issue from the fix that 24266 addresses.

I did spend some time last week and thought the patch sets that upgrade fabric8 which had fixes that might plausibly explain the problem seen by @redsk  .

I could revert the one line patch that you wanted to revert that was requested by @holdenk but I feel caught in the middle and would prefer that you both agree.

**Comment by jkleckner** [comment]
I reverted the SparkR per your instruction and also rebased to branch-3.0.

**Comment by dongjoon-hyun** [comment]
Thanks, @jkleckner . For the SparkR part, we can ignore it in this PR if it fails again. And, `disabling` should be handled by a separate PR if we do it in `branch-3.0`.
> I reverted the SparkR per your instruction and also rebased to branch-3.0.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, all!

Merged to branch-3.0 for Apache Spark 3.0.2.

I also hit the same issue, `ExecutorPodsWatchSnapshotSource`, reported by @redsk without this patch on the latest `branch-3.0`. I agree that is another watcher issue.

**Comment by dongjoon-hyun** [comment]
@redsk Could you file a new JIRA for your observation, please?

**Comment by jkleckner** [comment]
@dongjoon-hyun  Thank you for the merge and hanging in there.  Hopefully @shockdm will revive the 2.4 branch fix kicked off in #29496

**Comment by redsk** [comment]
@dongjoon-hyun I've created [SPARK-33349](https://issues.apache.org/jira/browse/SPARK-33349) as requested. Thanks

**Comment by dongjoon-hyun** [comment]
Thank you so much, @redsk !

**Pull request closed without merging**
